File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.cpp

#include <bits/stdc++.h>
using namespace std;


vector<string> solution(vector<string> players, vector<string> callings) {
    vector<string> answer;
    map<string, int> m;

    for(int i = 0; i<players.size(); i++){
        m[players[i]] = i;
    }

    for(auto c : callings){
        int current_pos = m[c];
        string prev_player = players[current_pos - 1];
        swap(players[current_pos], players[current_pos - 1]);

        m[c]--;
        m[prev_player]++;
    }

    return players;
}

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.js
function solution(players, callings) {
  var answer = [];
  let m = new Map();
  for(let i = 0; i<players.length; i++){
    m.set(players[i], i);
  }

  for(let c of callings){
    let cIdx = m.get(c); 
    let prevIdx = cIdx - 1;
    let prevPlayer = players[prevIdx];
    m.set(c, prevIdx);
    m.set(prevPlayer, cIdx);

    players[prevIdx] = c;
    players[cIdx] = prevPlayer;
  }

  return players;
}